sql/src/main/java/io/crate/executor/transport/ExecutionNodesTask.java

```diff
@@ -0,0 +1,304 @@
+import io.crate.planner.node.dql.MergeNode;
+import io.crate.types.DataTypes;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.breaker.CircuitBreaker;
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.logging.Loggers;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import javax.annotation.Nonnull;
+import java.util.*;
+
+
+public class ExecutionNodesTask extends JobTask {
+
+    private static final ESLogger LOGGER = Loggers.getLogger(ExecutionNodesTask.class);
+
+    private final TransportJobAction transportJobAction;
+    private final List<ListenableFuture<TaskResult>> results;
+    private final boolean hasDirectResponse;
+    private final SettableFuture<TaskResult> result;
+    private final PageDownstreamContext pageDownstreamContext;
+    private final Streamer<?>[] streamers;
+    private final StatsTables statsTables;
+    private final UUID mergeOperationId;
+    private final Map<String, Collection<ExecutionNode>> nodesByServer;
+
+    protected ExecutionNodesTask(UUID jobId,
+                                 JobContextService jobContextService,
+                                 PageDownstreamFactory pageDownstreamFactory,
+                                 final StatsTables statsTables,
+                                 ThreadPool threadPool,
+                                 TransportJobAction transportJobAction,
+                                 TransportCloseContextNodeAction transportCloseContextNodeAction,
+                                 CircuitBreaker circuitBreaker,
+                                 MergeNode mergeNode,
+                                 ExecutionNode... executionNodes) {
+        super(jobId);
+        this.statsTables = statsTables;
+        mergeOperationId = UUID.randomUUID();
+        this.transportJobAction = transportJobAction;
+
+        String ramAccountingContextId = String.format("%s: %s", mergeNode.name(), mergeOperationId);
+        final RamAccountingContext ramAccountingContext = new RamAccountingContext(ramAccountingContextId, circuitBreaker);
+
+        result = SettableFuture.create();
+        results = ImmutableList.<ListenableFuture<TaskResult>>of(result);
+        hasDirectResponse = hasDirectResponse(executionNodes);
+
+        RowDownstream rowDownstream = new QueryResultRowDownstream(result, jobId, mergeNode.executionNodeId(), jobContextService);
+        PageDownstream finalMergePageDownstream = pageDownstreamFactory.createMergeNodePageDownstream(
+                mergeNode,
+                rowDownstream,
+                ramAccountingContext,
+                Optional.of(threadPool.executor(ThreadPool.Names.SEARCH))
+        );
+
+        streamers = DataTypes.getStreamer(mergeNode.inputTypes());
+        pageDownstreamContext = new PageDownstreamContext(
+                finalMergePageDownstream,
+                streamers,
+                executionNodes[executionNodes.length -1 ].executionNodes().size()
```

Comment:
imho we should think about this naming, we have now 2 executionNodes defining different things... thought about that before, didn't found a proper name for one of this, lets discuss or make a proposal ;)